### PR TITLE
VACMS-13362: Adds migration for NCA Facilities API data.

### LINKED ISSUE #13362: Migrate additional NCA fields from Facilities API to CMS
## User Story or Problem Statement

Per #13123 , map the NCA fields from the Facilities API in Lighthouse to the CMS fields created in #13340 

## Acceptance Criteria
- [ ] I expect if a facility is recognized as missing during the migration, it should auto archive
- [x] I expect that the following fields are mapped as shown:

| Drupal field | Facilities API Selector | Notes |
| ------------ | ---------------------- | ----- |
| field_nca_classification | properties/classification | new text (plain) field |
| field_link/uri | properties/website | uri of existing Link field |
| field_link/title | facility name | title of existing Link field |
| field_timezone | properties/time_zone | new text (plain) field |
| field_mailing_address | properties/address/mailing | new Address field |
| field_address | properties/address/physical | new Address field |
| field_fax_number | properties/phone/fax | existing telephone field |
| field_phone_number | properties/phone/main | existing telephone field |
| field_office_hours | properties/hours | existing node field |

### More complex imported fields

<details><summary>field_link</summary>

```
source:
  fields:
    -
      name: field_link
      label: 'Link'

process:
  field_link/uri:
    -
      plugin: str_replace
      source: field_link
      search: 'NULL'
      replace: ''
  field_linkl/title:
    -
      plugin: str_replace
      source: <properties/name> TODO: Clarify this title
      search: 'NULL'
      replace: ''
```

</details>

<details><summary>field_mailing_address</summary>

```
source:
  fields:
    -
      name: city
      label: city
      selector: properties/address/mailing/city
    -
      name: state
      label: state
      selector: properties/address/mailing/state
    -
      name: zip
      label: zip
      selector: properties/address/mailing/zip
    -
      name: address1
      label: address1
      selector: properties/address/mailing/address_1
    -
      name: address2
      label: address2
      selector: properties/address/mailing/address_2
    -
      name: address3
      label: address3
      selector: properties/address/mailing/address_3
```

Refer to work done in 
va_gov_migrate/config/install/migrate_plus.migration.va_node_facility_vet_centers.yml
for more details

Combine API lines address_2 & address_3 into Drupal address line 2

</details>

<details><summary>field_address</summary>

```
source:
  fields:
    -
      name: city
      label: city
      selector: properties/address/physical/city
    -
      name: state
      label: state
      selector: properties/address/physical/state
    -
      name: zip
      label: zip
      selector: properties/address/physical/zip
    -
      name: address1
      label: address1
      selector: properties/address/physical/address_1
    -
      name: address2
      label: address2
      selector: properties/address/physical/address_2
    -
      name: address3
      label: address3
      selector: properties/address/physical/address_3
```

Refer to work done in 
va_gov_migrate/config/install/migrate_plus.migration.va_node_facility_vet_centers.yml
for more details

Combine API lines address_2 & address_3 into Drupal address line 2

</details>

<details><summary>field_phone_number</summary>

```
source:
  fields:
    -
      name: phone-main
      label: phone-main
      selector: properties/phone/main
process:
  field_phone_number:
    -
      plugin: str_replace
      source: phone-main
      search: Ext.
      replace: x
    -
      plugin: explode
      delimiter: x
    -
      plugin: callback
      callable: trim
    -
      plugin: concat
      delimiter: ' x'
```

Refer to work done in 
va_gov_migrate/config/install/migrate_plus.migration.va_node_facility_vet_centers.yml
for more details

</details>

<details><summary>field_office_hours</summary>

```
source:
  fields:
    -
      name: hours
      label: hours
      selector: properties/hours
process:
  field_office_hours:
    plugin: va_field_office_hours
    source: hours
```

Refer to work done in 
va_gov_migrate/config/install/migrate_plus.migration.va_node_facility_vet_centers.yml
for more details

</details>

- [ ] I expect that migration of NCA content results in the CMS fields being correctly populated.
- [ ] I expect that any Facilities product documentation related to NCA data has been updated.

## Implementation steps
See above

## Design principles

Veteran-centered
- [ ] `Single source of truth`: Increase reliability and consistency of content on VA.gov by providing a single source of truth.
- [ ] `Accessible, plain language`: Provide guardrails and guidelines to ensure content quality.
- [ ] `Purposely structured content`: Ensure Content API can deliver content whose meaning matches its structure.
- [ ] `Content lifecycle governance`: Produce tools, processes and policies to maintain content quality throughout its lifecycle.

Editor-centered
- [ ] `Purpose-driven`: Create an opportunity to involve the editor community in VA’s mission and content strategy goals.
- [ ] `Efficient`: Remove distractions and create clear, straightforward paths to get the job done.
- [ ] `Approachable`: Offer friendly guidance over authoritative instruction.
- [ ] `Consistent`: Reduce user’s mental load by allowing them to fall back on pattern recognition to complete tasks.
- [ ] `Empowering`: Provide clear information to help editors make decisions about their work.

### Team
Please check the team(s) that will do this work.

- [x] `Facilities`

### PR DESCRIPTION
**Pull request by omahane**
## Description

Closes #13362 

## Testing done
Manually

## Screenshots
Screen Shot 2023-05-31 at 7 46 10 AM
Screen Shot 2023-05-31 at 7 46 49 AM

### Test migration
- As an admin, go to [National Cemetery of the Alleghenies](https://pr13858-nlnnnfn5vedzn3oopc5rv5gxwhsm2mmz.ci.cms.va.gov/node/4595/edit) and note what data exists currently
- In a separate window, go to [Baxter Springs City Soldiers' Lot](https://pr13858-nlnnnfn5vedzn3oopc5rv5gxwhsm2mmz.ci.cms.va.gov/node/4432/edit) and note what data exists currently
- In a separate window, go to [Benicia Arsenal Post Cemetery](https://pr13858-nlnnnfn5vedzn3oopc5rv5gxwhsm2mmz.ci.cms.va.gov/node/27552/edit) and note what data exists currently
- In a separate window, run the [NCA migration](https://pr13858-nlnnnfn5vedzn3oopc5rv5gxwhsm2mmz.ci.cms.va.gov/admin/structure/migrate/manage/facility/migrations/va_node_facility_nca/execute) (making sure to have Update checked).
- [ ] Validate that the data in the NCA facilities in the open windows are now updated.

### Test archiving
- Open a PHP terminal in the PR Tugboat instance 
- Run the `drush` command: `drush va_gov_migrate:flag-missing-facilities`
- Does it only archive NCA facilities?
  - [ ] Validate that [Test Cemetery 1](https://pr13858-nlnnnfn5vedzn3oopc5rv5gxwhsm2mmz.ci.cms.va.gov/nca-facilities/locations/test-cemetery-1) was archived
  - [ ] Validate that the revision log message is associated with CMS Migrator
  - [ ] Validate that the revision log message is accurate
  - [ ] Validate that [Test Cemetery 2](https://pr13858-nlnnnfn5vedzn3oopc5rv5gxwhsm2mmz.ci.cms.va.gov/nca-facilities/locations/test-cemetery-2) was archived
  - [ ] Validate that the revision log message is associated with CMS Migrator
  - [ ] Validate that the revision log message is accurate
- Does it flag non-NCA facilities?
  - [ ] Validate that [Test VBA 1](https://pr13858-nlnnnfn5vedzn3oopc5rv5gxwhsm2mmz.ci.cms.va.gov/continental/locations/test-vba-1) was flagged 
  - [ ] Validate that the revision log message is associated with CMS Migrator
  - [ ] Validate that the revision log message is accurate

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `CMS Team`
- [ ] `Public websites`
- [x] `Facilities`
- [ ] `User support`

### CODE REVIEW

File: docroot/modules/custom/va_gov_migrate/src/Commands/Commands.php [hunk 1]
```diff
@@ -155,30 +155,69 @@ public function flagMissingFacilities() {
...
+  /**
+   * Archive a facility.
+   *
+   * @param \Drupal\node\NodeInterface $facility
+   *   The facility to archive.
+   */
+  protected function archiveRemovedFacility(NodeInterface $facility) {
+    $facility->set('moderation_state', 'archived');
+    $facility->setRevisionLogMessage('Archived due to removal from Facility API.');
+    $facility->save();
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_migrate/src/Commands/Commands.php:220]: I am not sure if this creates a new revision or just alters the current.  I am not able to test this right now to know for sure, but please check the revision log display shows this as a new revision attributed to the CMS Migrator.

I think we want to include some other things
```suggestion
    $facility->setNewRevision(TRUE);
    $facility->setUnpublished();
    Assign to CMS Migrator user.
    $facility->setRevisionUserId(1317);
    // Prevents some other actions.
    $facility->setSyncing(TRUE);
    $facility->setChangedTime(time());
    $facility->isDefaultRevision(TRUE);
    $facility->setRevisionCreationTime(time());
    $facility->save();
```

**Review by swirtSJW — CHANGES_REQUESTED** [review]
The migration looks great.  I added some suggested changes on the archival part.

File: docroot/modules/custom/va_gov_migrate/src/Commands/Commands.php [hunk 2]
```diff
@@ -155,30 +155,69 @@ public function flagMissingFacilities() {
       $this->migrateChannelLogger->log(LogLevel::WARNING, 'The facility API returned %count facilities, which seems suspicious. Flagging aborted.', $vars);
       return;
     }
+    $facilities_to_archive = ['nca_facility'];
```
[the remark this answers is not in the document]
Comment (omahane) [reply docroot/modules/custom/va_gov_migrate/src/Commands/Commands.php:158]: Good idea.

**Comment by omahane** [comment]
@swirtSJW Thank you for the review while on vacation, no less. I have pulled those suggestions in and added QA steps. I'll get someone else to give it a thorough going over.

**Review by swirtSJW — COMMENTED** [review]
Two thoughts.  On about facility ops, the other about setting facility status to closed and clearing out status morinfo.

File: docroot/modules/custom/va_gov_migrate/src/Commands/Commands.php [hunk 3]
```diff
@@ -155,30 +155,78 @@ public function flagMissingFacilities() {
       $this->migrateChannelLogger->log(LogLevel::WARNING, 'The facility API returned %count facilities, which seems suspicious. Flagging aborted.', $vars);
       return;
     }
+    $facility_bundles_to_archive = ['nca_facility'];
```
[the remark this answers is not in the document]
Comment (omahane) [reply docroot/modules/custom/va_gov_migrate/src/Commands/Commands.php:158]: These thoughts led me to some spelunking, and I saw this in `va_gov_facilities/src/EventSubscriber/FacilitiesSubscriber.php`, which could be refactored to use `isFacilityWithStatus`:
```
 protected function clearNormalStatusDetails(EntityInterface $entity): void {
    if ($entity instanceof NodeInterface) {
      $facilitiesWithStatus = [
        'health_care_local_facility',
        'nca_facility',
        'vba_facility',
        'vet_center',
        'vet_center_outstation',
      ];
```
Beyond this scope of this work, but thought I'd let you know, @swirtSJW

**Comment by omahane** [comment]
> To avoid confusion from conflicting data, this should set status to closed, and clear out the status morinfo.

I am unclear what the conflicting data to which you are referring might be. I can see that there may be information in the more info field even if we set it to "closed". Is the idea that we wouldn't want any data in the more info field that would be at odds with the "closed" status. Just trying to confirm. I am going to bring it up in refinement today to get Dave C/Michelle's feedback.

**Comment by omahane** [comment]
As a follow up, we talked about this change to the ACs in refinement (changing the status and removing the status more info content) and decided to keep those changes in another ticket (if it's decided to do it at all).

**Review by dsasser — APPROVED** [review]

**Comment by swirtSJW** [comment]
> > To avoid confusion from conflicting data, this should set status to closed, and clear out the status morinfo.
> 
> I am unclear what the conflicting data to which you are referring might be. I can see that there may be information in the more info field even if we set it to "closed". Is the idea that we wouldn't want any data in the more info field that would be at odds with the "closed" status. Just trying to confirm. I am going to bring it up in refinement today to get Dave C/Michelle's feedback.

Answered here  https://github.com/department-of-veterans-affairs/va.gov-cms/issues/13362#issuecomment-1580816072

File: docroot/modules/custom/va_gov_migrate/src/Commands/Commands.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (swirtSJW) [reply docroot/modules/custom/va_gov_migrate/src/Commands/Commands.php]: I created a ticket based on this finding https://github.com/department-of-veterans-affairs/va.gov-cms/issues/14045

**Pull request merged**